cmd/el-proxy/main.go

```diff
@@ -0,0 +1,482 @@
+	}
+	if cfg.Metrics && cfg.PProf && mAddr == pAddr {
+		return fmt.Errorf("metrics and pprof cannot be enabled on the same address:port: %s", mAddr)
+	}
+	if cfg.Metrics {
+		go metrics.CollectProcessMetrics(time.Second)
+		exp.Setup(fmt.Sprintf("%v:%v", cfg.MetricsServer.Addr, cfg.MetricsServer.Port))
+	}
+	if cfg.PProf {
+		genericconf.StartPprof(pAddr)
+	}
+	return nil
+}
+
+type ExpressLaneProxy struct {
+	stopwaiter.StopWaiter
+	config              *ExpressLaneProxyConfig
+	roundTimingInfo     timeboost.RoundTimingInfo
+	auctionContractAddr common.Address
+	expressLaneTracker  *gethexec.ExpressLaneTracker
+	dataSignerFunc      signature.DataSignerFunc
+
+	seqNumAssignmentMutex sync.Mutex
+	lastRoundNumber       uint64
+	nextSequenceNumber    uint64
+}
+
+type HeaderProviderAdapter struct {
+	*ethclient.Client
+}
+
+func (a *HeaderProviderAdapter) HeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Header, error) {
+	return a.Client.HeaderByNumber(ctx, big.NewInt((int64)(number)))
+}
+
+func NewExpressLaneProxy(
+	ctx context.Context,
+	config *ExpressLaneProxyConfig,
+	stack *node.Node,
+) (*ExpressLaneProxy, error) {
+	client, err := GetClientFromURL(ctx, config.RPCURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	arbClient := ethclient.NewClient(client)
+
+	if !common.IsHexAddress(config.AuctionContractAddress) {
+		return nil, fmt.Errorf("invalid auction-contract-address \"%v\"", config.AuctionContractAddress)
+	}
+	auctionContractAddr := common.HexToAddress(config.AuctionContractAddress)
+	auctionContract, err := express_lane_auctiongen.NewExpressLaneAuction(auctionContractAddr, arbClient)
+	if err != nil {
+		return nil, err
+	}
+
+	roundTimingInfo, err := gethexec.GetRoundTimingInfo(auctionContract)
+	if err != nil {
+		return nil, err
+	}
+
+	expressLaneTracker := gethexec.NewExpressLaneTracker(
+		*roundTimingInfo,
+		time.Millisecond*250,
+		&HeaderProviderAdapter{arbClient},
+		auctionContract,
+		auctionContractAddr,
+		&params.ChainConfig{ChainID: big.NewInt(config.ChainId)},
+		0,
+	)
+
+	_, dataSignerFunc, err := util.OpenWallet("el-proxy", &config.Wallet, big.NewInt(config.ChainId))
+	if err != nil {
+		return nil, fmt.Errorf("error opening wallet: %w", err)
+	}
+
+	elProxy := &ExpressLaneProxy{
+		config:              config,
+		roundTimingInfo:     *roundTimingInfo,
+		auctionContractAddr: auctionContractAddr,
+		expressLaneTracker:  expressLaneTracker,
+		dataSignerFunc:      dataSignerFunc,
+		lastRoundNumber:     math.MaxUint64,
+	}
+
+	elAPIs := []rpc.API{{
+		Namespace: "eth",
+		Version:   "1.0",
+		Service:   elProxy,
+		Public:    true,
+	}}
+
+	stack.RegisterAPIs(elAPIs)
+	return elProxy, nil
+}
+
+func (p *ExpressLaneProxy) Start(ctx context.Context) {
+	p.StopWaiterSafe.Start(ctx, p)
+	p.expressLaneTracker.Start(ctx)
+}
+
+func (p *ExpressLaneProxy) StopAndWait() {
+	p.StopWaiterSafe.StopAndWait()
+	p.expressLaneTracker.StopAndWait()
```

Comment:
Should we have the linter ignore the errcheck errors, or should we handle them somehow?

Reply:
I should've used StopWaiter here which panics on err. Fixed.